plugins/openvr/src/ViveControllerManager.cpp

```diff
@@ -262,6 +428,311 @@ bool ViveControllerManager::isHeadControllerMounted() const {
+        return pose;
+    }
+}
+
+
+void ViveControllerManager::trackFinger(int hand, int jointIndex1, int jointIndex2, int jointIndex3, int jointIndex4,
+                                        controller::StandardPoseChannel joint1, controller::StandardPoseChannel joint2,
+                                        controller::StandardPoseChannel joint3, controller::StandardPoseChannel joint4) {
+
+    glm::vec3 point1 = getRollingAverageHandPoint(hand, jointIndex1);
+    glm::vec3 point2 = getRollingAverageHandPoint(hand, jointIndex2);
+    glm::vec3 point3 = getRollingAverageHandPoint(hand, jointIndex3);
+    glm::vec3 point4 = getRollingAverageHandPoint(hand, jointIndex4);
+
+    glm::vec3 wristPos = getRollingAverageHandPoint(hand, 0);
+    glm::vec3 thumb2 = getRollingAverageHandPoint(hand, 2);
+    glm::vec3 pinkie1 = getRollingAverageHandPoint(hand, 17);
+
+    // 1st
+    glm::vec3 palmFacing = glm::normalize(glm::cross(pinkie1 - wristPos, thumb2 - wristPos));
+    glm::vec3 handForward = glm::normalize(point1 - wristPos);
+    glm::vec3 x = glm::normalize(glm::cross(palmFacing, handForward));
+    glm::vec3 y = glm::normalize(point2 - point1);
+    glm::vec3 z = (hand == 0) ? glm::cross(y, x) : glm::cross(x, y);
+    glm::mat4 rotMat1 = glm::mat4(glm::vec4(x, 0.0f),
+                                  glm::vec4(y, 0.0f),
+                                  glm::vec4(z, 0.0f),
+                                  glm::vec4(0.0f, 0.0f, 0.0f, 1.0f));
+    glm::quat rot1 = glm::normalize(glmExtractRotation(rotMat1));
+    if (!isNaN(rot1)) {
+        _inputDevice->_poseStateMap[joint1] = controller::Pose(point1, rot1);
+    }
+
+
+    // 2nd
+    glm::vec3 x2 = x; // glm::normalize(glm::cross(point3 - point2, point2 - point1));
+    glm::vec3 y2 = glm::normalize(point3 - point2);
+    glm::vec3 z2 = (hand == 0) ? glm::cross(y2, x2) : glm::cross(x2, y2);
+
+    glm::mat4 rotMat2 = glm::mat4(glm::vec4(x2, 0.0f),
+                                  glm::vec4(y2, 0.0f),
+                                  glm::vec4(z2, 0.0f),
+                                  glm::vec4(0.0f, 0.0f, 0.0f, 1.0f));
+    glm::quat rot2 = glm::normalize(glmExtractRotation(rotMat2));
+    if (!isNaN(rot2)) {
+        _inputDevice->_poseStateMap[joint2] = controller::Pose(point2, rot2);
+    }
+
+
+    // 3rd
+    glm::vec3 x3 = x; // glm::normalize(glm::cross(point4 - point3, point3 - point1));
+    glm::vec3 y3 = glm::normalize(point4 - point3);
+    glm::vec3 z3 = (hand == 0) ? glm::cross(y3, x3) : glm::cross(x3, y3);
+
+    glm::mat4 rotMat3 = glm::mat4(glm::vec4(x3, 0.0f),
+                                  glm::vec4(y3, 0.0f),
+                                  glm::vec4(z3, 0.0f),
+                                  glm::vec4(0.0f, 0.0f, 0.0f, 1.0f));
+    glm::quat rot3 = glm::normalize(glmExtractRotation(rotMat3));
+    if (!isNaN(rot3)) {
+        _inputDevice->_poseStateMap[joint3] = controller::Pose(point3, rot3);
+    }
+
+
+    // 4th
+    glm::quat rot4 = rot3;
```

Comment:
what are x2, x3. and rot4?  why aren't they different?

Reply:
x2 and x3 = x because finger joints are hinges.
rot4 = rot3 because rot4 is the orientatiion of the tip of the finger, the same as the orientation of joint 3.